contracts/pcv/sentinel/PCVSentinel.sol

```diff
@@ -0,0 +1,173 @@
+
+    // The set of all guards
+    EnumerableSet.AddressSet private guards;
+
+    constructor(
+        address _core
+    ) CoreRef(_core) {
+        _setContractAdminRole(keccak256("PCV_SENTINEL_ADMIN_ROLE"));
+    }
+
+    // ---------- Read-Only API ----------
+
+    /**
+     * @notice returns whether or not the given address is a guard
+     * @param guard the address to check
+     */
+    function isGuard(address guard) 
+        external 
+        view
+        override
+        returns (bool) 
+    {
+        return guards.contains(guard);
+    }
+
+    /**
+     * @notice returns a list of all guards
+     */
+    function allGuards() 
+        external
+        view
+        override
+        returns (address[] memory all) 
+    {
+        all = new address[](guards.length());
+
+        for (uint i = 0; i < guards.length(); i++) {
+            all[i] = guards.at(i);
+        }
+
+        return all;
+    }
+
+    /**
+     * @notice returns the list of all guards that pass their check
+     * @dev you *can* use this on-chain if you want, but this is made
+     * to be called off-chain to pass into protecManys and not hyper-efficient
+     */
+    function checkAll()
+        external 
+        view
+        override 
+        returns (address[] memory) 
+    {
+        uint256 numGuardsCanProtec;
+
+        for (uint256 i=0; i<guards.length(); i++) {
+            if(IGuard(guards.at(i)).check()) {
+                numGuardsCanProtec++;
+            }
+        }
+
+        address[] memory guardsWhoCanProtec = new address[](numGuardsCanProtec);
+
+        for (uint256 i=0; i<guards.length(); i++) {
+            if(IGuard(guards.at(i)).check()) {
+                numGuardsCanProtec--;
+                guardsWhoCanProtec[numGuardsCanProtec] = guards.at(i);
+            }
+        }
+
+        return guardsWhoCanProtec;
+    }
+
+    // ---------- Governor-or-Admin-Or-Guardian-Only State-Changing API ----------
+
+    /**
+     * @notice adds a guard contract to the PCV Sentinel
+     * @param guard the guard-contract to add
+     */
+    function knight(address guard) 
+        external 
+        override 
+        isGovernorOrGuardianOrAdmin
+    {
+        guards.add(guard);
+
+        // Inform the kingdom of this glorious news
+        emit GuardAdded(guard);
+    }
+
+    /**
+     * @notice removes a guard
+     * @param traitor the guard-contract to remove
+     */
+    function slay(address traitor) 
+        external 
+        override 
+        isGovernorOrGuardianOrAdmin
+    {
+        guards.remove(traitor);
+
+        // Inform the kingdom of this sudden and inevitable betrayal
+        emit GuardRemoved(traitor);
+    }
+
+    // ---------- Public State-Changing API ----------
+
+    /**
+     * @notice checks the guard and runs its protec actions if needed
+     * @param guard the guard to activate
+     */
+    function protec(address guard) 
+        external
+        override
```

Comment:
this function should have a reentrancy guard

Reply:
Now that we're not using delegatecall, and we're ensuring that we aren't calling ourselves with the address(this) check, no re-entrancy guard is needed

Reply:
I would rather be safe than sorry, especially when executing arbitrary calldata against targets. We should add a reentrancy guard here.

Reply:
There is literally no way that re-entrancy can happen. We're good.